plugins/scale/scale-title-overlay.cpp

```diff
@@ -0,0 +1,452 @@
+    {
+        pad = {0, 0, 0, 0};
+        if (!should_have_overlay())
+        {
+            return;
+        }
+
+        if (overlay.tex.tex == (GLuint) - 1)
+        {
+            render_overlay();
+        }
+
+        switch (pos)
+        {
+          case position::CENTER:
+        {
+            int w = (int)std::ceil(overlay.tex.width / par.output_scale);
+            int h = (int)std::ceil(overlay.tex.height / par.output_scale);
+
+            /* If we are rendering on top of a dialog, then w, h can be larger
+             * than the size of our box (but will be within the parent's box).
+             * In this case, we have to add to our padding */
+            if (w > box.width)
+            {
+                pad.left = pad.right = (w - box.width + 1) / 2;
+            }
+
+            if (h > box.height)
+            {
+                pad.top = pad.bottom = (h - box.height + 1) / 2;
+            }
+        }
+        break;
+
+          case position::BOTTOM:
+            pad.bottom = (int)std::ceil(overlay.tex.height / par.output_scale);
+            break;
+
+          case position::TOP:
+            pad.top = (int)std::ceil(overlay.tex.height / par.output_scale);
+            break;
+        }
+    }
+};
+
+scale_show_title_t::scale_show_title_t() :
+    view_filter{[this] (auto)
+    {
+        update_title_overlay_opt();
+    }},
+
+    scale_end{[this] (wf::signal_data_t*)
+    {
+        show_view_title_overlay = title_overlay_t::NEVER;
+        last_title_overlay = nullptr;
+        mouse_update.disconnect();
+    }
+},
+
+add_title_overlay{[this] (wf::signal_data_t *data)
+    {
+        const std::string& opt = show_view_title_overlay_opt;
+        if (opt == "never")
+        {
+            /* TODO: support changing this option while scale is running! */
+            return;
+        }
+
+        const std::string& pos_opt = title_position;
+        view_title_overlay_t::position pos = view_title_overlay_t::position::CENTER;
+        if (pos_opt == "top")
+        {
+            pos = view_title_overlay_t::position::TOP;
+        } else if (pos_opt == "bottom")
+        {
+            pos = view_title_overlay_t::position::BOTTOM;
+        }
+
+        auto signal = static_cast<scale_transformer_added_signal*>(data);
+        auto tr     = signal->transformer;
+        auto view   = tr->get_transformed_view();
+        auto ol     = std::make_shared<view_title_overlay_t>(*tr, title_font_size,
+            bg_color, text_color, pos);
+
+        auto cb = std::make_shared<wf::scale_transformer_t::render_callback_t>(
+            [this, ol, view] (const wf::framebuffer_t& fb,
+                              const wf::region_t& damage)
+        {
+            if (show_view_title_overlay == title_overlay_t::NEVER)
+            {
+                return;
+            }
+
+            if (show_view_title_overlay == title_overlay_t::MOUSE)
+            {
+                auto tmp_view = view;
```

Comment:
seems to me we should have a helper function `get_toplevel_view` which returns the topmost parent.

Reply:
Yes, that makes sense; I've copied the one from [view.cpp](https://github.com/WayfireWM/wayfire/blob/27a5a97bec803c70b175ae6d0f84a2925c5e329a/src/view/view.cpp#L73)